libraries/standard/http/src/http_client.c

```diff
@@ -29,6 +34,91 @@ static HTTPStatus_t _sendHttpBody( const HTTPTransportInterface_t * pTransport,
                                    const uint8_t * pRequestBodyBuf,
                                    size_t reqBodyBufLen );
 
+/**
+ * @brief Write header based on parameters. This method also adds a trailing "\r\n".
+ * If a trailing "\r\n" already exists in the HTTP header, this method backtracks
+ * in order to write over it and updates the length accordingly.
+ *
+ * @param pRequestHeaders Request header buffer information.
+ * @param pField The header field name to write.
+ * @param fieldLen The byte length of the header field name.
+ * @param pValue The header value to write.
+ * @param valueLen The byte length of the header field value.
+ *
+ * @return #HTTP_SUCCESS if successful. If there was insufficient memory in the
+ * application buffer, then #HTTP_INSUFFICIENT_MEMORY is returned.
+ */
+static HTTPStatus_t _addHeader( HTTPRequestHeaders_t * pRequestHeaders,
+                                const char * pField,
+                                size_t fieldLen,
+                                const char * pValue,
+                                size_t valueLen );
+
+/*-----------------------------------------------------------*/
+
+static HTTPStatus_t _addHeader( HTTPRequestHeaders_t * pRequestHeaders,
+                                const char * pField,
+                                size_t fieldLen,
+                                const char * pValue,
+                                size_t valueLen )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    uint8_t * pBufferCur = pRequestHeaders->pBuffer + pRequestHeaders->headersLen;
+    size_t toAddLen = 0;
+    uint8_t hasTrailingLine = 0u;
+
+    /* Backtrack before trailing "\r\n" (HTTP header end) if it's already written.
+     * Note that this method also writes trailing "\r\n" before returning. */
+    if( strncmp( ( char * ) pBufferCur - ( 2 * HTTP_HEADER_LINE_SEPARATOR_LEN ),
+                 "\r\n\r\n", 2 * HTTP_HEADER_LINE_SEPARATOR_LEN ) == 0 )
+    {
+        /* Set this flag to backtrack in case of HTTP_INSUFFICIENT_MEMORY. */
+        hasTrailingLine = 1u;
+        pBufferCur -= HTTP_HEADER_LINE_SEPARATOR_LEN;
+        pRequestHeaders->headersLen -= HTTP_HEADER_LINE_SEPARATOR_LEN;
+    }
+
+    /* Check if there is enough space in buffer for additional header. */
+    toAddLen = fieldLen + HTTP_HEADER_FIELD_SEPARATOR_LEN + valueLen + \
+               HTTP_HEADER_LINE_SEPARATOR_LEN +                        \
+               HTTP_HEADER_LINE_SEPARATOR_LEN;
+
+    /* If we have enough room for the new header line, then write it to the header buffer. */
+    if( ( pRequestHeaders->headersLen + toAddLen ) <= pRequestHeaders->bufferLen )
+    {
+        /* Write "Field: Value \r\n\r\n" to headers. */
+        memcpy( pBufferCur, pField, fieldLen );
```

Comment:
Have you considered using `sprintf` here instead of `memcpy`? (I have checked with Richard that we are fine using `sprintf` for now as we can possibly employ underlying optimization strategies later)

Reply:
I would actually prefer that. @sarenameas?

Reply:
Nevermind, I don't think that `sprintf` is a candidate for use here as the `fieldLen` and `valueLen` parameters cannot be passed to it, and the user is not required to pass NULL-terminated strings